course5/homework10/linked_list.h

```diff
@@ -0,0 +1,221 @@
+        }
+        else
+            std::cout << "copy -- no next" << std::endl;
+        mSize = src.mSize;
+        std::cout << "copy -- set size" << std::endl;
+    }
+    // Move construction
+    LinkedList(LinkedList&& src)
+    {
+        std::cout << "Move construction: from " << src.front() << std::endl;
+        data = std::move(src.data);
+        next = std::move(src.next);
+        mSize = src.mSize;
+    }
+    // Copy assignment
+    LinkedList& operator=(const LinkedList& src)
+    {
+        std::cout << "Copy assignment: from " << src.front() << std::endl;
+        data = std::make_unique<T>(src.front());
+        if (src.next)
+            next = std::make_unique<LinkedList<T>>(*src.next.get());
+        mSize = src.mSize;
+        return *this;
+    }
+    // Move assignment
+    LinkedList& operator=(LinkedList&& src)
+    {
+        std::cout << "Move assignment: from " << src.front() << std::endl;
+        data = std::move(src.data);
+        next = std::move(src.next);
+        mSize = src.mSize;
+        return *this;
+    }
+
+    bool operator==(LinkedList& cmp) const
+    {
+        std::cout << mSize << " " << cmp.mSize << std::endl;
+        if (mSize != cmp.mSize)
+            return false;
+        if (data != cmp.data)
+            return false;
+        if (!(next && cmp.next))
+             return true;
+        if  (next != cmp.next)
+            return false;
+        return true;
+    }
+
+    int size() const { return mSize; }
+    T front() const // TODO should this be a pointer? const?
+    {
+        if (data)
+            return *data.get();
+        throw std::range_error("Calling front on empty LinkedList is undefined.");
+    }
+    void push_front(const T& val)
+    {
+        if (mSize > 0)
+        {
+            LinkedList<T> new_node;
+            new_node.data = std::move(this->data);
+            new_node.next = std::move(this->next);
+            data = std::make_unique<T>(val);
+            next = std::make_unique<LinkedList<T>>(std::move(new_node));
+        }
+        else
+        {
+            data = std::make_unique<T>(val);
+        }
+        mSize++;
+    }
+
+    template< class... Args >
+    void emplace_front(Args&&... args)
+    {
+        if (mSize > 0)
+        {
+            LinkedList<T> new_node;
+            new_node.data = std::move(this->data);
+            new_node.next = std::move(this->next);
+            data = std::make_unique<T>(args...);
+            next = std::make_unique<LinkedList<T>>(std::move(new_node));
+        }
+        else
+        {
+            data = std::make_unique<T>(args...);
+        }
+        mSize++;
+    }
+
+    void pop_front()
+    {
+        if (next)
+        {
+            LinkedList<T> new_head = std::move(*next.get());
+            data = std::move(new_head.data);
+            next = std::move(new_head.next);
+            mSize--;
+        }
+    }
+
+    void print() const
```

Comment:
And I'd like to move to iterators here, but I'd also like to understand what's going wrong.